In [2]:
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.metrics import classification_report,confusion_matrix, accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier, MLPRegressor
from sklearn.preprocessing import StandardScaler

In [3]:
data = pd.read_csv("telecom_churn_data.csv", header = None,names = ["State", "Account_Length", "Area_Code", "Phone","International_Plan","VMail_Plan", "VMail_Message","Day_Mins","Day_Calls","Day_Charge","Eve_Mins","Eve_Calls","Eve_Charge","Night_Mins","Night_Calls","Night_Charge","International_Mins","International_Calls","International_Charge","CustServ_calls","Churn"])
data.head()

State  Account_Length  Area_Code      Phone International_Plan VMail_Plan  \
0    KS             128        415   382-4657                 no        yes   
1    OH             107        415   371-7191                 no        yes   
2    NJ             137        415   358-1921                 no         no   
3    OH              84        408   375-9999                yes         no   
4    OK              75        415   330-6626                yes         no   

   VMail_Message  Day_Mins  Day_Calls  Day_Charge   ...     Eve_Calls  \
0             25     265.1        110       45.07   ...            99   
1             26     161.6        123       27.47   ...           103   
2              0     243.4        114       41.38   ...           110   
3              0     299.4         71       50.90   ...            88   
4              0     166.7        113       28.34   ...           122   

   Eve_Charge  Night_Mins  Night_Calls  Night_Charge  International_Mins  \
0       16.78       244.7           91         11.01                10.0   
1       16.62       254.4          103         11.45                13.7   
2       10.30       162.6          104          7.32                12.2   
3        5.26       196.9           89          8.86                 6.6   
4       12.61       186.9          121          8.41                10.1   

   International_Calls  International_Charge  CustServ_calls    Churn  
0                    3                  2.70               1   False.  
1                    3                  3.70               1   False.  
2                    5                  3.29               0   False.  
3                    7                  1.78               2   False.  
4                    3                  2.73               3   False.  

[5 rows x 21 columns]

In [4]:
data.Churn[data.Churn == ' True.'] = int(1)
data.International_Plan[data.International_Plan == ' yes'] = int(1)
data.VMail_Plan[data.VMail_Plan == ' yes'] = int(1)

C:\Users\PranaviPadegal\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\PranaviPadegal\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\PranaviPadegal\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing import

In [6]:
data.Churn[data.Churn == ' False.'] = int(0)
data.International_Plan[data.International_Plan == ' no'] = int(0)
data.VMail_Plan[data.VMail_Plan == ' no'] = int(0)

C:\Users\PranaviPadegal\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\PranaviPadegal\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\PranaviPadegal\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing import

In [9]:
data.International_Plan = data.International_Plan.astype('int')
data.VMail_Plan = data.VMail_Plan.astype('int')
data.dtypes

State                    object
Account_Length            int64
Area_Code                 int64
Phone                    object
International_Plan        int32
VMail_Plan                int32
VMail_Message             int64
Day_Mins                float64
Day_Calls                 int64
Day_Charge              float64
Eve_Mins                float64
Eve_Calls                 int64
Eve_Charge              float64
Night_Mins              float64
Night_Calls               int64
Night_Charge            float64
International_Mins      float64
International_Calls       int64
International_Charge    float64
CustServ_calls            int64
Churn                    object
dtype: object

In [11]:
y = data.Churn.astype('int')
x = data.loc[:,["Area_Code","International_Plan","VMail_Plan", "VMail_Message","Day_Mins","Day_Calls","Day_Charge","Eve_Mins","Eve_Calls","Eve_Charge","Night_Mins","Night_Calls","Night_Charge","International_Mins","International_Calls","International_Charge","CustServ_calls"]]

In [13]:
scaler = StandardScaler()

In [15]:
x = scaler.fit_transform(x)

In [16]:
x

array([[-0.52139206, -0.32699969,  1.65391562, ..., -0.58348728,
        -0.10200542, -0.43388538],
       [-0.52139206, -0.32699969,  1.65391562, ..., -0.58348728,
         1.24148062, -0.43388538],
       [-0.52139206, -0.32699969, -0.60462577, ...,  0.23039815,
         0.69065134, -1.19906826],
       ...,
       [-0.68694075, -0.32699969,  1.65391562, ..., -0.58348728,
         0.98621827, -0.43388538],
       [-0.52139206, -0.32699969,  1.65391562, ..., -0.58348728,
         1.45643839, -1.19906826],
       [-0.52139206, -0.32699969, -0.60462577, ..., -0.17654456,
         0.66378162,  0.3312975 ]])

In [25]:
x_train, x_test,y_train,y_test = train_test_split(x,y,test_size = 0.3,random_state = 20)

In [26]:
x_train.shape

(3231, 17)

### Building a model using MLP Classifier with 3 layers each having 100 Perceptro

In [40]:
model_mlp = MLPClassifier(hidden_layer_sizes=(100,100,100),batch_size=10,learning_rate_init=0.01,max_iter=2000,random_state=10)
model_mlp.fit(x_train,y_train)

MLPClassifier(activation='relu', alpha=0.0001, batch_size=10, beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100, 100, 100), learning_rate='constant',
       learning_rate_init=0.01, max_iter=2000, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=10, shuffle=True,
       solver='adam', tol=0.0001, validation_fraction=0.1, verbose=False,
       warm_start=False)

In [41]:
y_predict = model_mlp.predict(x_test)

1. Out of "1175+27" not churn values, "1175" are predicted as not churn and "27" are predicted as churn.
2. Out of "47+137" churn values, "47" are predicted as not churn and "137" are Predicted as churn which is far better than the earlier models like svm and Logistic regression.

In [42]:
confusion_matrix(y_test,y_predict)

array([[1175,   27],
       [  47,  137]], dtype=int64)

### Accuracy of 94.6 % is observed which is quite good.

In [43]:
accuracy_score(y_test,y_predict)

0.9466089466089466

In [44]:
print(classification_report(y_test,y_predict))

             precision    recall  f1-score   support

          0       0.96      0.98      0.97      1202
          1       0.84      0.74      0.79       184

avg / total       0.94      0.95      0.95      1386



In [49]:
from sklearn.linear_model import Perceptron

In [65]:
perceptron = Perceptron(alpha = 0.01,shuffle = True, random_state = 30)

In [66]:
perceptron.fit(x_train, y_train)
y_predict = model_mlp.predict(x_test)

C:\Users\PranaviPadegal\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.perceptron.Perceptron'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


In [67]:
confusion_matrix(y_test, y_predict)

array([[1175,   27],
       [  47,  137]], dtype=int64)

In [64]:
accuracy_score(y_test,y_predict)

0.9466089466089466